torch/csrc/autograd/python_variable.cpp

```diff
@@ -1941,24 +1942,55 @@ static PyObject* THPVariable_NewWithVar(
     PyTypeObject* type,
     Variable _var,
     c10::impl::PyInterpreterStatus status) {
-  // This function overwrite the Tensor's pyobj field without extra checks
-  // Make sure it is not set otherwise we would leak memory
-  auto mb_obj = _var.unsafeGetTensorImpl()->pyobj_slot()->check_pyobj(
-      self_interpreter.get());
-  TORCH_CHECK(
-      !mb_obj.has_value() || !mb_obj.value(),
-      "Creating a new Tensor subclass ",
-      type->tp_name,
-      " but the raw Tensor object is already associated to a python object ",
-      "of type ",
-      mb_obj.value()->ob_type->tp_name);
-
   // Make sure that the reinterpret into a THPVariable* will be valid
   TORCH_CHECK(
       PyType_IsSubtype(type, &THPVariableType),
       "Creating a Tensor subclass from a class ",
       "that does not inherit from Tensor is not possible. Make sure your class inherits from Tensor.");
 
+  // This function overwrite the Tensor's pyobj field without extra checks
+  // Make sure it is not set otherwise we would leak memory
+  auto mb_obj = _var.unsafeGetTensorImpl()->pyobj_slot()->check_pyobj(
+      self_interpreter.get());
+
+  // Under some circumstances, we may attempt to create a new Python
+  // object for a variable that already has a Python object.  The most common
+  // situation this can occur is if you have a TorchDispatchMode active that
+  // is returning a subclass from lift_fresh (which is invoked to
+  // appropriately "wrap" a constant tensor into whatever ambient modes are
+  // active.)
+  //
+  // In general, it is impossible to handle this case compositionally.
+  // Suppose you have a user call ATensor(...) when a mode is active that
+  // is transforming all ops to output BTensor, where ATensor and BTensor
+  // are completely unrelated subclasses and there is no way to compose them.
```

Comment:
In this case, I would expect that ATensor can be constructed based on a regular Tensor? And the mode BTensor can also handle that?

I would expect that the call that happens when you do `t = ATensor(factory())` under BTensor mode is:
- `raw_t = factory()` is called.
- The mode wraps the result as `b_t = BTensor(raw_t)`
- The subclass wraps the result as `t = ATensor(b_t)`.

Reply:
This seems pretty complicated.

When I say `ATensor(factory())`, irregardless of the details of how `ATensor.__new__` is implemented, this desugars into the following calls:

```
t = factory()
r = ATensor.__new__(t)
r.__init__(t)
```

TorchDispatchMode can interpose on the internal implementations of each of these individual calls, but what it cannot do is change this sequence of three calls. Furthermore, while in some situations the `factory()` call will be interposable, others are not; for example, I could call `ATensor([1, 2, 3])`; clearly TorchDispatchMode cannot change the meaning of `[1, 2, 3]`. So realistically, I only actually get to interpose on the new/init calls.

Under your proposal, you want this sequence to be equivalent to `ATensor(BTensor(...))` in the end. In this case, this would have to happen entirely inside of `ATensor.__new__`. Previously, this would be implemented via something like the following pseudocode:

```
class _TensorBase:
    def __new__(cls, tensor_arg):
        tensor_cdata = base_tensor_ctor(tensor_arg)  # this is TorchDispatchMode overrideable
        return _create_tensor_cls_with_cpp_tensor(cls, tensor_cdata)

class Tensor(_TensorBase):
    pass

class A(Tensor):
    def __new__(cls, *a_args, **a_kwargs):
         ...
         return super().__new__(cls, tensor)  # or maybe a direct as_subclass or make_subclass call, which indirectly call _TensorBase.__new__
```

So there is an immediate problem, which is that the control is inverted from what you suggest. A.__new__ calls TensorBase.__new__, which calls base_tensor_ctor which produces the BTensor. Ordinarily, we wouldn't "re"-invoke the A.__new__ constructor at this point; we're already there!

Let's suppose, somehow, you workaround this problem. There's another problem: there is no fixed contract for `A.__new__` arguments. For example, FakeTensor takes four required arguments; in general there may not be a single argument constructor that only does wrapping (this is a generally useful concept that we can strongly advise subclasses support, indeed, arguably `aten.lift` is precisely this operation), but there is nothing forcing this contract to be upheld, and no currently codified way to tell if a given subclass constructor implements this protocol.

Now, A could manually call `base_tensor_ctor` somehow (not exactly sure how to do this) and arrange to handle the case where this returns something that's not a regular Tensor. But that's tantamount to just making A/B compositional. Which we should probably tell people how to do, but it's a userspace change, not an infra level change.

Reply:
`but what it cannot do is change this sequence of three calls` - what is preventing us from calling torch.tensor, and then lift ?

Reply:
I read this comment but it's still not clear to me why can't call `torch.tensor` and then lift